### PR TITLE
creation new template  deploy vmss and custom script extenstion with master slave architecture ( updated with Github contribution guide and best practice ARM)

### PR DESCRIPTION
**Pull request by cloudmelon**
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use resourceGroup().location for resource locations
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/bp-checklist.md

- [ ] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

*
*
*

### CODE REVIEW

File: 201-vmss-master-slave-customscript/README.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Deploy a VM Scale Set of Linux VMs with a custom script extension in master / slave architecture
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/README.md]: add the buttons for deploy and visualization

File: 201-vmss-master-slave-customscript/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,492 @@
+﻿{
+  "$schema": "http://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json",
+  "contentVersion": "1.0.0.0",
+  "parameters": {
+    "vmssName": {
+      "type": "String",
+      "defaultValue": "cloudmelonvmss",
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/azuredeploy.json]: use a globally unique defaultValue via uniqueString  - add minLength & maxLength as appropriate

File: 201-vmss-master-slave-customscript/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+      "metadata": {
+        "description": "String used as a base for naming resources. Must be 3-10 characters in length and globally unique across Azure Stack. A hash is prepended to this string for some resources, and resource-specific information is appended."
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/azuredeploy.json]: Azure Stack - VMSS is not supported on Azure Stack yet

File: 201-vmss-master-slave-customscript/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+      }
+    },
+    "masterNodeCount": {
+      "type": "int",
+      "maxValue": 50,
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/azuredeploy.json]: is <=0 valid?

File: 201-vmss-master-slave-customscript/azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+    "dataNodeSku": {
+      "defaultValue": "Standard_D4S_v3",
+      "type": "String",
+      "metadata": {
+        "description": "Size of VMs in the VM Scale Set for data node."
+      }
+    },
+    "dataNodeCount": {
+      "type": "int",
+      "maxValue": 100,
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/azuredeploy.json]: is <=0 valid?

File: 201-vmss-master-slave-customscript/azuredeploy.json [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+      "metadata": {
+        "description": "Number of data nodes in VMSS; if singlePlacementGroup is true (the default), then this value must be 100 or less; if singlePlacementGroup is false, then ths value must be 100 or less"
+      }
+    },      
+    "adminUsername": {
+      "type": "String",
+      "defaultValue": "cloudmelonadmin",
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/azuredeploy.json]: remove defaultValue

File: 201-vmss-master-slave-customscript/azuredeploy.json [hunk 7]
```diff
@@ -0,0 +1,492 @@
...
+      "metadata": {
+        "description": "Admin username on all VMs."
+      }
+    },
+    "adminPassword": {
+      "type": "SecureString",
+      "defaultValue": "cloudmelonadmin@2018",
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/azuredeploy.json]: remove defaultValue

File: 201-vmss-master-slave-customscript/azuredeploy.json [hunk 8]
```diff
@@ -0,0 +1,492 @@
...
+    "osImageSku": {
+      "type": "string",
+      "defaultValue": "7.4",
+      "metadata": {
+        "description": "The Ubuntu version for the VM. This will pick a fully patched image of this given Ubuntu version. Default value: 14.04.3-LTS"
+      }
+    }
+  },
+  "variables": {
+    "location": "[resourceGroup().location]",
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/azuredeploy.json]: this must be a parameter

File: 201-vmss-master-slave-customscript/azuredeploy.json [hunk 9]
```diff
@@ -0,0 +1,492 @@
...
+    "masterNodeLoadBalancerProbeName": "[concat('LBHttpProbe', parameters('vmssName'))]",
+    "dataNodeLoadBalancerProbeName": "[concat('LBHttpProbe', parameters('vmssName'))]",
+    "masterNodeLoadBalancerNatPoolName": "[concat('LBNatPool', parameters('vmssName'))]",
+    "dataNodeLoadBalancerNatPoolName": "[concat('LBNatPool', parameters('vmssName'))]"
+  },
+  "resources": [
+    {
+      "type": "Microsoft.Storage/storageAccounts",
+      "name": "[variables('storageAccountName')]",
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/azuredeploy.json]: update apiVersion

File: 201-vmss-master-slave-customscript/azuredeploy.json [hunk 10]
```diff
@@ -0,0 +1,492 @@
...
+      "location": "[variables('location')]",
+      "properties": {
+        "accountType": "[variables('storageAccountType')]"
+      }
+    },
+    {
+      "apiVersion": "2017-04-01",
+      "type": "Microsoft.Network/virtualNetworks",
+      "name": "[variables('vnetName')]",
+      "location": "[resourceGroup().location]",
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/azuredeploy.json]: all location properties must be [parameters('location')]

File: 201-vmss-master-slave-customscript/azuredeploy.json [hunk 11]
```diff
@@ -0,0 +1,492 @@
...
+              "addressPrefix": "10.0.1.0/24"
+            }
+          }
+        ]
+      }
+    },
+    {
+      "type": "Microsoft.Network/publicIPAddresses",
+      "name": "[variables('masterPublicIPAddressName')]",
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/azuredeploy.json]: update apiVersion

File: 201-vmss-master-slave-customscript/azuredeploy.json [hunk 12]
```diff
@@ -0,0 +1,492 @@
...
+        "publicIPAllocationMethod": "Dynamic",
+        "dnsSettings": {
+          "domainNameLabel": "[concat(variables('vmssDomainName'),'-master-node')]"
+        }
+      }
+    },
+    {
+      "type": "Microsoft.Network/publicIPAddresses",
+      "name": "[variables('dataPublicIPAddressName')]",
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/azuredeploy.json]: update apiVersion

File: 201-vmss-master-slave-customscript/azuredeploy.json [hunk 13]
```diff
@@ -0,0 +1,492 @@
...
+        "publicIPAllocationMethod": "Dynamic",
+        "dnsSettings": {
+          "domainNameLabel": "[concat(variables('vmssDomainName'),'-data-node')]"
+        }
+      }
+    },
+    {
+      "type": "Microsoft.Network/loadBalancers",
+      "name": "[variables('masterNodeLoadBalancerName')]",
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/azuredeploy.json]: update apiVersion

File: 201-vmss-master-slave-customscript/azuredeploy.json [hunk 14]
```diff
@@ -0,0 +1,492 @@
...
+        ]
+      },
+      "dependsOn": [
+        "[concat('Microsoft.Network/publicIPAddresses/', variables('masterPublicIPAddressName'))]"
+      ]
+    },
+    {
+      "type": "Microsoft.Network/loadBalancers",
+      "name": "[variables('dataNodeLoadBalancerName')]",
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/azuredeploy.json]: update apiVersion

File: 201-vmss-master-slave-customscript/azuredeploy.json [hunk 15]
```diff
@@ -0,0 +1,492 @@
...
+    },
+    {
+      "type": "Microsoft.Compute/virtualMachineScaleSets",
+      "sku": {
+        "name": "[parameters('masterNodeSku')]",
+        "tier": "Standard",
+        "capacity": "[parameters('masterNodeCount')]"
+      },
+      "name": "[concat(parameters('vmssName'), '-master-node')]",
+      "apiVersion": "2016-03-30",
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/azuredeploy.json]: update apiVersion

File: 201-vmss-master-slave-customscript/azuredeploy.json [hunk 16]
```diff
@@ -0,0 +1,492 @@
...
+      "location": "[variables('location')]",
+      "properties": {
+        "upgradePolicy": {
+          "mode": "Manual"
+        },
+        "virtualMachineProfile": {
+          "storageProfile": {
+            "osDisk": {
+              "vhdContainers": [
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/azuredeploy.json]: use an implicit managed disk

File: 201-vmss-master-slave-customscript/azuredeploy.json [hunk 17]
```diff
@@ -0,0 +1,492 @@
...
+              {
+                "type": "Microsoft.Compute/virtualMachines/extensions",
+                "name": "[concat(parameters('vmssName'),'-LinuxCustomScriptExtension')]",
+                "properties": {
+                  "publisher": "Microsoft.OSTCExtensions",
+                  "type": "CustomScriptForLinux",
+                  "typeHandlerVersion": "1.3",
+                  "autoUpgradeMinorVersion": true,
+                  "settings": {
+                    "fileUris": [ "https://cloudmelongithub.blob.core.windows.net/cloudmelonarmmasterslave/testcloudmelonscript01.sh" ],
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/azuredeploy.json]: See [BP guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#deployment-artifacts-nested-templates-scripts) for how to handle artifacts

File: 201-vmss-master-slave-customscript/azuredeploy.json [hunk 18]
```diff
@@ -0,0 +1,492 @@
...
+      "apiVersion": "2016-03-30",
+      "location": "[variables('location')]",
+      "properties": {
+        "upgradePolicy": {
+          "mode": "Manual"
+        },
+        "virtualMachineProfile": {
+          "storageProfile": {
+            "osDisk": {
+              "vhdContainers": [
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/azuredeploy.json]: use an implicit managed disk

File: 201-vmss-master-slave-customscript/azuredeploy.json [hunk 19]
```diff
@@ -0,0 +1,492 @@
...
+              {
+                "type": "Microsoft.Compute/virtualMachines/extensions",
+                "name": "[concat(parameters('vmssName'),'-LinuxCustomScriptExtension')]",
+                "properties": {
+                  "publisher": "Microsoft.OSTCExtensions",
+                  "type": "CustomScriptForLinux",
+                  "typeHandlerVersion": "1.3",
+                  "autoUpgradeMinorVersion": true,
+                  "settings": {
+                    "fileUris": [ "https://cloudmelongithub.blob.core.windows.net/cloudmelonarmmasterslave/testcloudmelonscript02.sh" ],
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/azuredeploy.json]: See [BP guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#deployment-artifacts-nested-templates-scripts) for how to handle artifacts

File: 201-vmss-master-slave-customscript/azuredeploy.parameters.json [hunk 20]
```diff
@@ -0,0 +1,36 @@
+{
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/azuredeploy.parameters.json]: remove parameters that have defaultValue, use GEN values for all others

File: 201-vmss-master-slave-customscript/azuredeploy.parameters.json [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+	"$schema": "http://schema.management.azure.com/schemas/2015-01-01/deploymentParameters.json#",
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/azuredeploy.parameters.json]: https

File: 201-vmss-master-slave-customscript/metadata.json [hunk 22]
```diff
@@ -0,0 +1,8 @@
+{
+  "$schema": "http://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json",
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/metadata.json]: Must be:   "$schema": "https://aka.ms/azure-quickstart-templates-metadata-schema#",

File: 201-vmss-master-slave-customscript/azuredeploy.json [hunk 23]
```diff
@@ -0,0 +1,492 @@
+﻿{
+  "$schema": "http://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json",
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/azuredeploy.json]: https

File: 201-vmss-master-slave-customscript/azuredeploy.json (hunk 3 above)
Comment (cloudmelon) [reply 201-vmss-master-slave-customscript/azuredeploy.json]: updated description.

File: 201-vmss-master-slave-customscript/azuredeploy.json (hunk 4 above)
Comment (cloudmelon) [reply 201-vmss-master-slave-customscript/azuredeploy.json]: update constraints

File: 201-vmss-master-slave-customscript/azuredeploy.json (hunk 5 above)
Comment (cloudmelon) [reply 201-vmss-master-slave-customscript/azuredeploy.json]: update constraints with minValue

File: 201-vmss-master-slave-customscript/azuredeploy.json (hunk 9 above)
Comment (cloudmelon) [reply 201-vmss-master-slave-customscript/azuredeploy.json]: updated api version

File: 201-vmss-master-slave-customscript/azuredeploy.parameters.json (hunk 20 above)
Comment (cloudmelon) [reply 201-vmss-master-slave-customscript/azuredeploy.parameters.json]: updated

**Comment by cloudmelon** [comment]
Updated everything and finally passed !!
image
@bmoore-msft  Thank you for your suggestions

**Comment by cloudmelon** [comment]
@bmoore-msft Recheck every single point what you have been addressed.  Thank you for your suggestions. Feel free to reach out to me in case of any queries. Little by little the bird makes its nest :)

**Comment by cloudmelon** [comment]
Hi @bmoore-msft  how are you ? Any other comments to improve the quality of my commits? Thanks :)

File: 201-vmss-master-slave-customscript/README.md [hunk 24]
```diff
@@ -0,0 +1,17 @@
...
+<a href="http://armviz.io/#/?load=https://github.com/cloudmelon/azure-quickstart-templates/blob/master/201-vmss-master-slave-customscript/azuredeploy.json" target="_blank">
+<img src="https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/1-CONTRIBUTION-GUIDE/images/visualizebutton.png"/>
+</a>
+
+
+## Parameter Restrictions
+======================
+
+vmssName must be 3-10 characters in length. If it isn't globally unique, it is possible that this template will still deploy properly, but we don't recommend relying on this pseudo-probabilistic behavior.
+instanceCount must be 20 or less. VM Scale Set supports upto 100 VMs and one should add more storage accounts to support this number.
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/README.md]: this doesn't seem to align with the template parameters - we need to make this more clear.

re: storageAccounts, doesn't look like we're using any

File: 201-vmss-master-slave-customscript/azuredeploy.json [hunk 25]
```diff
@@ -0,0 +1,471 @@
...
+    "dataNodeLoadBalancerProbeName": "[concat('LBHttpProbe', variables('vmssName'))]",
+    "masterNodeLoadBalancerNatPoolName": "[concat('LBNatPool', variables('vmssName'))]",
+    "dataNodeLoadBalancerNatPoolName": "[concat('LBNatPool', variables('vmssName'))]"
+  },
+  "resources": [
+    {
+      "type": "Microsoft.Network/virtualNetworks",
+      "name": "[variables('vnetName')]",
+      "apiVersion": "2017-06-01",
+      "location": "[resourceGroup().location]",
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/azuredeploy.json]: all location propeties need to use the parameter

File: 201-vmss-master-slave-customscript/azuredeploy.json [hunk 26]
```diff
@@ -0,0 +1,471 @@
...
+              {
+                "type": "Microsoft.Compute/virtualMachines/extensions",
+                "name": "[concat(variables('vmssName'),'-LinuxCustomScriptExtension')]",
+                "properties": {
+                  "publisher": "Microsoft.OSTCExtensions",
+                  "type": "CustomScriptForLinux",
+                  "typeHandlerVersion": "1.3",
+                  "autoUpgradeMinorVersion": true,
+                  "settings": {
+                    "fileUris": [ "https://github.com/cloudmelon/azure-quickstart-templates/blob/master/201-vmss-master-slave-customscript/scripts/testcloudmelonscript.sh" ],
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/azuredeploy.json]: See [BP guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#deployment-artifacts-nested-templates-scripts) for how to handle artifacts

File: 201-vmss-master-slave-customscript/azuredeploy.json [hunk 27]
```diff
@@ -0,0 +1,471 @@
...
+                  "ipConfigurations": [
+                    {
+                      "name": "ipconfig",
+                      "properties": {
+                        "subnet": {
+                          "id": "[resourceId('Microsoft.Network/virtualNetworks/subnets', variables('vnetName'), variables('subnetMaster'))]"
+                        },
+                        "loadBalancerBackendAddressPools": [
+                          {
+                            "id": "[concat('/subscriptions/', subscription().subscriptionId,'/resourceGroups/', resourceGroup().name, '/providers/Microsoft.Network/loadBalancers/', variables('masterNodeLoadBalancerName'), '/backendAddressPools/', variables('masterNodeLoadBalancerBackEndName'))]"
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/azuredeploy.json]: simplify:

resourceId('Microsoft.Network/loadBalancers/backendAddressPools', variables('masterNodeLoadBalancerName'), variables('masterNodeLoadBalancerBackEndName'))

Here and below

File: 201-vmss-master-slave-customscript/azuredeploy.json [hunk 28]
```diff
@@ -0,0 +1,471 @@
...
+                  "ipConfigurations": [
+                    {
+                      "name": "ipconfig",
+                      "properties": {
+                        "subnet": {
+                          "id": "[resourceId('Microsoft.Network/virtualNetworks/subnets', variables('vnetName'), variables('subnetData'))]"
+                        },
+                        "loadBalancerBackendAddressPools": [
+                          {
+                            "id": "[concat('/subscriptions/', subscription().subscriptionId,'/resourceGroups/', resourceGroup().name, '/providers/Microsoft.Network/loadBalancers/', variables('dataNodeLoadBalancerName'), '/backendAddressPools/', variables('dataNodeLoadBalancerBackEndName'))]"
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/azuredeploy.json]: simplify as above

File: 201-vmss-master-slave-customscript/azuredeploy.json [hunk 29]
```diff
@@ -0,0 +1,471 @@
...
+              {
+                "type": "Microsoft.Compute/virtualMachines/extensions",
+                "name": "[concat(variables('vmssName'),'-LinuxCustomScriptExtension')]",
+                "properties": {
+                  "publisher": "Microsoft.OSTCExtensions",
+                  "type": "CustomScriptForLinux",
+                  "typeHandlerVersion": "1.3",
+                  "autoUpgradeMinorVersion": true, 
+                  "settings": {
+                    "fileUris": [ "https://github.com/cloudmelon/azure-quickstart-templates/blob/master/201-vmss-master-slave-customscript/scripts/testcloudmelonscript.sh" ],
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/azuredeploy.json]: See [BP guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#deployment-artifacts-nested-templates-scripts) for how to handle artifacts

File: 201-vmss-master-slave-customscript/azuredeploy.parameters.json (hunk 20 above)
Comment (bmoore-msft) [reply 201-vmss-master-slave-customscript/azuredeploy.parameters.json]: still needed

File: 201-vmss-master-slave-customscript/azuredeploy.json [hunk 30]
```diff
@@ -0,0 +1,466 @@
...
+               {  
+                  "name":"[variables('masterNodeLoadBalancerBackendName')]"
+               }
+            ],
+            "loadBalancingRules":[  
+               {  
+                  "name":"roundRobinLBRule",
+                  "properties":{  
+                     "frontendIPConfiguration":{  
+                        "id":"[concat(resourceId('Microsoft.Network/loadBalancers', variables('masterNodeLoadBalancerName')), '/frontendIPConfigurations/', variables('masterNodeLoadBalancerFrontEndName'))]"
```
Comment (bmoore-msft) [inline 201-vmss-master-slave-customscript/azuredeploy.json]: simplify, just use resourceId() - just one comment here but all need to be updated.

**Comment by cloudmelon** [comment]
@bmoore-msft Update all the resource ID in a simpler way.

**Pull request merged**